compiler/optimizer/OMRValuePropagation.hpp

```diff
@@ -281,118 +284,32 @@ class ValuePropagation : public TR::Optimization
    ValueConstraint *copyValueConstraints(ValueConstraints &valueConstraints);
 
 #if USE_TREES
-   class ValueConstraintHandler : public TREE_HANDLER<ValueConstraint>
+   class ValueConstraintHandler : public TREE_HANDLER <ValueConstraint>
       {
       public:
-      void setVP(TR::ValuePropagation *vp) {_vp = vp;}
```

Comment:
I'm not proposing a solution yet, just trying to understand.  Do you only need the mapping to the `TR::` namespace in order to construct the proper VP object?  In all other cases, is passing around the base of the hierarchy (i.e., a `OMR::ValuePropagation` class) sufficient?

A bit of history...

At one point for extensible classes the "concrete class" in the TR namespace didn't exist and we simply used a typedef of the most derived class into the TR namespace.  However, this wasn't feasible for a few reasons, some of which may be relevant here, and especially if this is a general pattern that we will use as a precedent:

* typedefs cannot be forward declared,
* typedefs cannot "override" each other, so you have to manage it so that only one is visible to the compiler,
* typedefs of classes cannot be friended (actually they can in C++11, but not all build compilers support this feature yet)

The point about forward declaration may be the most relevant in this discussion and was the biggest deal-breaker for OMR.